src/transformers/modeling_longformer.py

```diff
@@ -0,0 +1,667 @@
+
+    def __init__(self, config):
+        super().__init__(config)
+
+        if isinstance(config.attention_window, int):
+            assert config.attention_window % 2 == 0, '`attention_window` has to be an even value'
+            assert config.attention_window > 0, '`attention_window` has to be positive'
+            config.attention_window = [config.attention_window] * config.num_hidden_layers  # one value per layer
+        else:
+            assert len(config.attention_window) == config.num_hidden_layers, (
+                '`len(attention_window)` should equal `num_hidden_layers`. '
+                f'Expected {config.num_hidden_layers}, given {len(config.attention_window)}'
+            )
+
+        if config.attention_mode == 'n2':
+            pass  # do nothing, use the default `modeling_bert.BertSelfAttention` (will OOM for long sequences)
+        elif config.attention_mode == 'longformer':
+            for i, layer in enumerate(self.encoder.layer):
+                # replace the `modeling_bert.BertSelfAttention` object with `LongformerSelfAttention`
+                layer.attention.self = LongformerSelfAttention(config, layer_id=i)
+        else:
+            raise ValueError(f'Expected values of `attention_mode` are "longformer" or "n2", given {config.attention_mode}')
+
+    def _pad_to_window_size(self, input_ids: torch.Tensor, attention_mask: torch.Tensor, token_type_ids: torch.Tensor,
+                            position_ids: torch.Tensor, inputs_embeds: torch.Tensor,
+                            attention_window: int, pad_token_id: int):
+        '''A helper function to pad tokens and mask to work with implementation of Longformer selfattention.'''
+
+        assert attention_window % 2 == 0
+        seqlen = input_ids.size(1)
+        padding_len = (attention_window - seqlen % attention_window) % attention_window
+        if padding_len > 0:
+            logger.info(
+                "Input ids are automatically padded from {} to {} to be a multiple of `config.attention_window`: {}".format(
+                    seqlen, seqlen + padding_len, attention_window
+                )
+            )
+            if input_ids is not None:
+                input_ids = F.pad(input_ids, (0, padding_len), value=pad_token_id)
+            if attention_mask is not None:
+                attention_mask = F.pad(attention_mask, (0, padding_len), value=False)  # no attention on the padding tokens
+            if token_type_ids is not None:
+                token_type_ids = F.pad(token_type_ids, (0, padding_len), value=0)  # pad with token_type_id = 0
+            if position_ids is not None:
+                # pad with position_id = pad_token_id as in modeling_roberta.RobertaEmbeddings
+                position_ids = F.pad(position_ids, (0, padding_len), value=pad_token_id)
+            if inputs_embeds is not None:
+                inputs_embeds = F.pad(inputs_embeds, (0, 0, 0, padding_len), value=0.0)  # pad with zero embeddings
```

Comment:
I think it's usually safer to concatenate the embeddings of the `pad_token_id` to the `input_embeds` in order to pad the `input_embeds`, so that we don't have to hard code the value 0.0 here.